src/controllers/publish.ts

```diff
@@ -104,27 +134,67 @@ export const uploadFile = async (req: Request, res: Response, next: NextFunction
     res.render('publish/upload', { revisit, errors });
 };
 
-export const importPreview = async (req: Request, res: Response, next: NextFunction) => {
-    const { dataset, revision, fileImport } = res.locals;
+// Special thanks ChatGPT...  The GovUK pagination algorithm
+function generateSequenceForNumber(highlight: number, end: number): (string | number)[] {
+    const sequence: (string | number)[] = [];
+
+    // Validate input
+    if (highlight > end) {
+        throw new Error(`Highlighted number must be between 1 and ${end}.`);
+    }
+
+    // Numbers before the highlighted number
+    if (highlight - 1 > 1) {
+        sequence.push(1, '...');
+        sequence.push(highlight - 1);
+    } else {
+        // eslint-disable-next-line @typescript-eslint/naming-convention
+        sequence.push(...Array.from({ length: highlight - 1 }, (_, index) => index + 1));
+    }
+
+    // Highlighted number
+    sequence.push(highlight);
+
+    // Numbers after the highlighted number
+    if (highlight + 1 < end) {
+        sequence.push(highlight + 1);
+        sequence.push('...', end);
+    } else {
+        // eslint-disable-next-line @typescript-eslint/naming-convention
+        sequence.push(...Array.from({ length: end - highlight }, (_, index) => highlight + 1 + index));
+    }
+
+    return sequence;
+}
```

Comment:
Can we stick this in `/src/utils/pagination` please? Just thinking it might be useful for routes outside the publish journey.

Reply:
I'll land this PR and raise a small one to extract and reuse.